GTEST/UCP/AMO: test wireup + amo + flush

evgeny-leksikov opened the pull request:
## What
Test if atomic operation is completed after ep flush which is in wireup state.

## Why ?
Found an uncertain behavior during SW AMO with loopback implementation.

## How ?
add new unit gtest

@yosefe pls take a look, is this a bug or expected behavior? for rcx and dcx (rare) flush completion callback is called before AMO is done

mellanox-github commented:
<details><summary>Mellanox CI: <b>FAILED</b> on 13 of 25 workers (click for details)</summary>

**Note:** the logs will be deleted after 20-Dec-2019

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/45db2b8dc6548351e3214ffb7cb42051/raw/a0b7b8862c8b608c4238ade20609bb1e447237cb/_main) | :x: FAILURE
[hpc-arm-cavium-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/45db2b8dc6548351e3214ffb7cb42051/raw/a82ce5f25b136b6c6948e1ded61dc1d603452866/hpc-arm-cavium-jenkins_W1) | :x: FAILURE
[hpc-arm-cavium-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/45db2b8dc6548351e3214ffb7cb42051/raw/2f03e15ad07d8d0e2a996fcf61febd5d5014b4dd/hpc-arm-cavium-jenkins_W2) | :x: FAILURE
[hpc-arm-hwi-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/45db2b8dc6548351e3214ffb7cb42051/raw/a4e28dffc130ea8664ad5e5c633f11bdaac6d035/hpc-arm-hwi-jenkins_W1) | :x: FAILURE
[hpc-arm-hwi-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/45db2b8dc6548351e3214ffb7cb42051/raw/4764aecddd21ac07be927a1dbf2960c25e496340/hpc-arm-hwi-jenkins_W2) | :x: FAILURE
[hpc-test-node-gpu_W0](https://gist.githubusercontent.com/mellanox-github/45db2b8dc6548351e3214ffb7cb42051/raw/92b92af712d15b0542f893c5a88c6991b232b7e6/hpc-test-node-gpu_W0) | :x: FAILURE
[hpc-test-node-gpu_W1](https://gist.githubusercontent.com/mellanox-github/45db2b8dc6548351e3214ffb7cb42051/raw/a2cd9947508b09766270d52606874019b7ab675c/hpc-test-node-gpu_W1) | :x: FAILURE
[hpc-test-node-legacy_W0](https://gist.githubusercontent.com/mellanox-github/45db2b8dc6548351e3214ffb7cb42051/raw/6f498bdab96e6e4d46624076cee35b3590c546d5/hpc-test-node-legacy_W0) | :x: FAILURE
[hpc-test-node-legacy_W1](https://gist.githubusercontent.com/mellanox-github/45db2b8dc6548351e3214ffb7cb42051/raw/65447e12f27ffaebf4879abc180474f396651132/hpc-test-node-legacy_W1) | :x: FAILURE
[hpc-test-node-new_W1](https://gist.githubusercontent.com/mellanox-github/45db2b8dc6548351e3214ffb7cb42051/raw/5494e21702589d13939b29501967831ea12c4d4e/hpc-test-node-new_W1) | :x: FAILURE
[hpc-test-node-new_W2](https://gist.githubusercontent.com/mellanox-github/45db2b8dc6548351e3214ffb7cb42051/raw/817fd21f8cc49117177ae82c2726647bcc748942/hpc-test-node-new_W2) | :x: FAILURE
[r-vmb-ppc-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/45db2b8dc6548351e3214ffb7cb42051/raw/c85ffc4ea3413b54d33a04b442cca47b720941cd/r-vmb-ppc-jenkins_W0) | :x: FAILURE
[r-vmb-ppc-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/45db2b8dc6548351e3214ffb7cb42051/raw/15f1916ac1a452a63d03382d07afa5f3846f15ff/r-vmb-ppc-jenkins_W1) | :x: FAILURE
[hpc-arm-cavium-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/45db2b8dc6548351e3214ffb7cb42051/raw/342186fbfa316617e4c731d8bbbb95a632c36b80/hpc-arm-cavium-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/45db2b8dc6548351e3214ffb7cb42051/raw/70240d4fc3b26f08bb4b18a85614946be02c0d7d/hpc-arm-cavium-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/45db2b8dc6548351e3214ffb7cb42051/raw/383c44b8a4f10c8fdd6c0e3fc874f84a009a509e/hpc-arm-hwi-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/45db2b8dc6548351e3214ffb7cb42051/raw/de0b0fe8ffe46871d17d1e19236ca2f5a2f4eeef/hpc-arm-hwi-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W2](https://gist.githubusercontent.com/mellanox-github/45db2b8dc6548351e3214ffb7cb42051/raw/10b5aa3528ecedde29ac99886e0ebacfaeca6ea1/hpc-test-node-gpu_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W3](https://gist.githubusercontent.com/mellanox-github/45db2b8dc6548351e3214ffb7cb42051/raw/5787a46239a53461e0f4d5089f5925dac1c8f202/hpc-test-node-gpu_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W2](https://gist.githubusercontent.com/mellanox-github/45db2b8dc6548351e3214ffb7cb42051/raw/b41bb128f787d48c3e5798b8eb61d9af49fc0862/hpc-test-node-legacy_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W3](https://gist.githubusercontent.com/mellanox-github/45db2b8dc6548351e3214ffb7cb42051/raw/945fbb54ddf68d4a122385b8949683bd210e4cd5/hpc-test-node-legacy_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W0](https://gist.githubusercontent.com/mellanox-github/45db2b8dc6548351e3214ffb7cb42051/raw/74307795478c52154101ee37681e8e77b5601497/hpc-test-node-new_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W3](https://gist.githubusercontent.com/mellanox-github/45db2b8dc6548351e3214ffb7cb42051/raw/d8e55d5a260978b8188830f21a7b56c3af5560e8/hpc-test-node-new_W3) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/45db2b8dc6548351e3214ffb7cb42051/raw/339e91bb530599904cdc469c12b7f3acc5068e6c/r-vmb-ppc-jenkins_W2) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/45db2b8dc6548351e3214ffb7cb42051/raw/522b5e2260e2268788e414bd84c1d9d0b3a9acc5/r-vmb-ppc-jenkins_W3) | :heavy_check_mark: SUCCESS

yosefe commented:
@evgeny-leksikov it happens because IB send completion does not really guarantee memory write
it's same error @hoopoepg saw with openshmem and multi-rail

evgeny-leksikov commented:
@yosefe ok this is why values are not matched but the problem that I saw with rc is amo was not performed before flush completion at all and if destroy the rkey in the callback it's not valid during amo execution. I guess there is some issue with wireup ep.

yosefe commented:
so can you modify the test to assert that UCP AMO request is completed (instead of checking value)?

evgeny-leksikov commented:
AMO returns status, not a request.

yosefe commented:
> AMO returns status, not a request.

so just remove value check?

evgeny-leksikov commented:
> > AMO returns status, not a request.
> 
> so just remove value check?

done

mellanox-github commented:
<details><summary>Mellanox CI: <b>FAILED</b> on 13 of 25 workers (click for details)</summary>

**Note:** the logs will be deleted after 22-Dec-2019

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/7bd296d112154ee0f4b7aa73d1a3f5eb/raw/7d445f83dc1959b3c8cdacb384966d1794b7a0f9/_main) | :x: FAILURE
[hpc-arm-cavium-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/7bd296d112154ee0f4b7aa73d1a3f5eb/raw/3042f33dd0b3352108fe5be18eee4dc6ce7890d7/hpc-arm-cavium-jenkins_W1) | :x: FAILURE
[hpc-arm-cavium-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/7bd296d112154ee0f4b7aa73d1a3f5eb/raw/6b0f9cd9eba17e9633e60ccb72418f5a58e0c16b/hpc-arm-cavium-jenkins_W2) | :x: FAILURE
[hpc-arm-hwi-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/7bd296d112154ee0f4b7aa73d1a3f5eb/raw/b6e36b7daf8d74fb3e48ff224b6df7325a8e5a6b/hpc-arm-hwi-jenkins_W1) | :x: FAILURE
[hpc-arm-hwi-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/7bd296d112154ee0f4b7aa73d1a3f5eb/raw/7ffc2ceedc130a19f1f30db743e7d8f27080958f/hpc-arm-hwi-jenkins_W2) | :x: FAILURE
[hpc-test-node-gpu_W0](https://gist.githubusercontent.com/mellanox-github/7bd296d112154ee0f4b7aa73d1a3f5eb/raw/1cde8663449a0c3f88ed9698fca4abb24ad51ae7/hpc-test-node-gpu_W0) | :x: FAILURE
[hpc-test-node-gpu_W1](https://gist.githubusercontent.com/mellanox-github/7bd296d112154ee0f4b7aa73d1a3f5eb/raw/6a32b62804c8de9b41fa3dbdae27419b95f55f6c/hpc-test-node-gpu_W1) | :x: FAILURE
[hpc-test-node-legacy_W0](https://gist.githubusercontent.com/mellanox-github/7bd296d112154ee0f4b7aa73d1a3f5eb/raw/ec2986f6ab20cab6d3399339171589d208f9e08d/hpc-test-node-legacy_W0) | :x: FAILURE
[hpc-test-node-legacy_W1](https://gist.githubusercontent.com/mellanox-github/7bd296d112154ee0f4b7aa73d1a3f5eb/raw/a2078c8d4d6630f478e5b11653fa947528617bba/hpc-test-node-legacy_W1) | :x: FAILURE
[hpc-test-node-new_W1](https://gist.githubusercontent.com/mellanox-github/7bd296d112154ee0f4b7aa73d1a3f5eb/raw/c392855abb1c9511e7c396aaf0f870da23584ec1/hpc-test-node-new_W1) | :x: FAILURE
[hpc-test-node-new_W2](https://gist.githubusercontent.com/mellanox-github/7bd296d112154ee0f4b7aa73d1a3f5eb/raw/117c3829c78b4f2733e002a3d623c5b778daa97e/hpc-test-node-new_W2) | :x: FAILURE
[r-vmb-ppc-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/7bd296d112154ee0f4b7aa73d1a3f5eb/raw/9423f22a2e625d210d63cf25c2828821719d392f/r-vmb-ppc-jenkins_W0) | :x: FAILURE
[r-vmb-ppc-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/7bd296d112154ee0f4b7aa73d1a3f5eb/raw/53fb4ab2d6e61ece78e3248b601431191ac297d9/r-vmb-ppc-jenkins_W1) | :x: FAILURE
[hpc-arm-cavium-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/7bd296d112154ee0f4b7aa73d1a3f5eb/raw/4da55c476b888f484df9166d7de2bb7a0efa14a9/hpc-arm-cavium-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/7bd296d112154ee0f4b7aa73d1a3f5eb/raw/4644b17ee6b923772a7b4390e417ccf1bbef3c30/hpc-arm-cavium-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/7bd296d112154ee0f4b7aa73d1a3f5eb/raw/6ba86945e87f7b21d2a080897ae5ad0f1a459dc9/hpc-arm-hwi-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/7bd296d112154ee0f4b7aa73d1a3f5eb/raw/bc8f7b83201f104e3df5116479426e592ecc150c/hpc-arm-hwi-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W2](https://gist.githubusercontent.com/mellanox-github/7bd296d112154ee0f4b7aa73d1a3f5eb/raw/f1dc3397c00a966159df34fc2c08a8f4983dc749/hpc-test-node-gpu_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W3](https://gist.githubusercontent.com/mellanox-github/7bd296d112154ee0f4b7aa73d1a3f5eb/raw/89049efa5c1ba126b308e9977b5dd45f8ef4d071/hpc-test-node-gpu_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W2](https://gist.githubusercontent.com/mellanox-github/7bd296d112154ee0f4b7aa73d1a3f5eb/raw/aa0a6dd72fbc70e572d0a97fdcc434b6792ec13e/hpc-test-node-legacy_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W3](https://gist.githubusercontent.com/mellanox-github/7bd296d112154ee0f4b7aa73d1a3f5eb/raw/97bd3ece738bec9e92b8ea69ec5ad94424b0e9c8/hpc-test-node-legacy_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W0](https://gist.githubusercontent.com/mellanox-github/7bd296d112154ee0f4b7aa73d1a3f5eb/raw/7a5233f792c8b7fa674de71b2745e75020622ac1/hpc-test-node-new_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W3](https://gist.githubusercontent.com/mellanox-github/7bd296d112154ee0f4b7aa73d1a3f5eb/raw/42d2bf08b4377b59b6df3a09058d9ee7222fff14/hpc-test-node-new_W3) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/7bd296d112154ee0f4b7aa73d1a3f5eb/raw/d32e66ec2c014f80fb285a4ccb427459eb1ef4a6/r-vmb-ppc-jenkins_W2) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/7bd296d112154ee0f4b7aa73d1a3f5eb/raw/5f58c222c0473c8ff1c39bfab523b8974def9415/r-vmb-ppc-jenkins_W3) | :heavy_check_mark: SUCCESS

dmitrygx commented on test/gtest/ucp/test_ucp_wireup.cc:
```diff
@@ -200,8 +219,36 @@ ucp_rkey_h test_ucp_wireup::get_rkey(ucp_ep_h ep, ucp_mem_h memh)
+    return (uint64_t)&m_recv_data[0];
+}
+
+
```
excessive empty line?

dmitrygx commented on test/gtest/ucp/test_ucp_wireup.cc:
```diff
@@ -1149,6 +1200,60 @@ class test_ucp_wireup_fallback_amo : public test_ucp_wireup {
+    }
+
+protected:
+    static void scomplete_cb(void *req, ucs_status_t status) {
```
`scomplete_cb` = send complete callback?

evgeny-leksikov replied:
yes, copy-pasted... this test is written for review to determine firstly if it's a valid usage. If so, then there is a bug in UCX.

evgeny-leksikov commented:
@yosefe I found a bug (sender/receiver memhs were mixed up) in the test and fixed it, now it seems working. Last question related to this. I added memhs cleanup to flush CB including target/destination memh. 
https://github.com/evgeny-leksikov/ucx/blob/091605f93a23fd00630f4bc8689de6c3d4d293ad/test/gtest/ucp/test_ucp_wireup.cc#L1232
Is it valid taking into account the fact that IB doesn't guarantee delivery to memory?

yosefe commented:
> Is it valid taking into account the fact that IB doesn't guarantee delivery to memory?

AFAIK yes

evgeny-leksikov commented:
Ok, then you can merge this PR after CI testing or close if you find it useless.

yosefe commented:
more testing is always good IMHO

mellanox-github commented:
<details><summary>Mellanox CI: <b>PASSED</b#  > on 25 workers (click for details)</summary>

**Note:** the logs will be deleted after 23-Dec-2019

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/93db855787cfccdcd7e1d36869aa9441/raw/15cd059b943d92cb639338bf8df5c575fd4e797d/_main) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/93db855787cfccdcd7e1d36869aa9441/raw/c8bce5edab8af6319d1ac3c104d29c56a359faf4/hpc-arm-cavium-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/93db855787cfccdcd7e1d36869aa9441/raw/917bd28dc0dbec0dabdac2279f6cdbd15bc8c1cb/hpc-arm-cavium-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/93db855787cfccdcd7e1d36869aa9441/raw/4a6568905d0d7950d92714867513b5952000f3f5/hpc-arm-cavium-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/93db855787cfccdcd7e1d36869aa9441/raw/1f79ddd6ea3ee980f6e05c0ee5219844f513e893/hpc-arm-cavium-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/93db855787cfccdcd7e1d36869aa9441/raw/8dbcbaf7d109dde613eee3ed9469915b06fe43ab/hpc-arm-hwi-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/93db855787cfccdcd7e1d36869aa9441/raw/c9357ddc5e6906f05ea43f2016ad69d296b90297/hpc-arm-hwi-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/93db855787cfccdcd7e1d36869aa9441/raw/a2d0bdfc8749b4f4a4c10b494d0e433a5a97ed00/hpc-arm-hwi-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/93db855787cfccdcd7e1d36869aa9441/raw/9b6e6ecc63da231cd2c27e5c9a245bc6b149a4d0/hpc-arm-hwi-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W0](https://gist.githubusercontent.com/mellanox-github/93db855787cfccdcd7e1d36869aa9441/raw/04ffd57fd1f98268c826a6a77ffd2bb285bfd752/hpc-test-node-gpu_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W1](https://gist.githubusercontent.com/mellanox-github/93db855787cfccdcd7e1d36869aa9441/raw/435f97f924a98354c3cbde3d87f67d8b852dd7f9/hpc-test-node-gpu_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W2](https://gist.githubusercontent.com/mellanox-github/93db855787cfccdcd7e1d36869aa9441/raw/9c2d6de11d0ded952a49832b39b5822cf98b4cdd/hpc-test-node-gpu_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W3](https://gist.githubusercontent.com/mellanox-github/93db855787cfccdcd7e1d36869aa9441/raw/c263d95f9eebcf0c4e427b16b0a66a4603932c12/hpc-test-node-gpu_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W0](https://gist.githubusercontent.com/mellanox-github/93db855787cfccdcd7e1d36869aa9441/raw/72a2db725436aa7115f9353019d5e3979c681dbf/hpc-test-node-legacy_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W1](https://gist.githubusercontent.com/mellanox-github/93db855787cfccdcd7e1d36869aa9441/raw/971903cfb351b7760b8cba7601b70635fce4a752/hpc-test-node-legacy_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W2](https://gist.githubusercontent.com/mellanox-github/93db855787cfccdcd7e1d36869aa9441/raw/f3026a62e4df5927e827a0e24dffc47d7b37c3fc/hpc-test-node-legacy_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W3](https://gist.githubusercontent.com/mellanox-github/93db855787cfccdcd7e1d36869aa9441/raw/9bac6b164e1dd4330e29cee50c17a69736fbbac8/hpc-test-node-legacy_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W0](https://gist.githubusercontent.com/mellanox-github/93db855787cfccdcd7e1d36869aa9441/raw/82738a0905345cc14690ce7e310355a87145f946/hpc-test-node-new_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W1](https://gist.githubusercontent.com/mellanox-github/93db855787cfccdcd7e1d36869aa9441/raw/ea0ebf854f384bd2a88d8877f90e21f720951f2f/hpc-test-node-new_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W2](https://gist.githubusercontent.com/mellanox-github/93db855787cfccdcd7e1d36869aa9441/raw/07d101a08b867cbfa056e6b810aea3df61888b92/hpc-test-node-new_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W3](https://gist.githubusercontent.com/mellanox-github/93db855787cfccdcd7e1d36869aa9441/raw/d1a68db4832a82c3094c4858d864e84745a784bf/hpc-test-node-new_W3) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/93db855787cfccdcd7e1d36869aa9441/raw/92009a410cc04953bd59ab7cfe35cdf2e4328f23/r-vmb-ppc-jenkins_W0) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/93db855787cfccdcd7e1d36869aa9441/raw/41278cf542fa3946e1bd8513998ca75e8debe9ef/r-vmb-ppc-jenkins_W1) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/93db855787cfccdcd7e1d36869aa9441/raw/b95e91a6295a828b206f1bf512f5196258ea50c3/r-vmb-ppc-jenkins_W2) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/93db855787cfccdcd7e1d36869aa9441/raw/685bc90fa5f715964d02876201f7c53c9eb97f6b/r-vmb-ppc-jenkins_W3) | :heavy_check_mark: SUCCESS

evgeny-leksikov commented:
@yosefe do you have other comments? Or I can squash before merge?

yosefe commented on test/gtest/ucp/test_ucp_wireup.cc:
```diff
@@ -1149,6 +1205,60 @@ class test_ucp_wireup_fallback_amo : public test_ucp_wireup {
+                     const std::string& test_case_name, const std::string& tls)
+    {
+        return enum_test_params_features(ctx_params, name, test_case_name, tls,
+                                         UCP_FEATURE_AMO32 | UCP_FEATURE_AMO64,
```
can we require only UCP_FEATURE_AMO32/UCP_FEATURE_AMO64, depending on sizeof(elem_type) ?

yosefe commented on test/gtest/ucp/test_ucp_wireup.cc:
```diff
@@ -1149,6 +1205,60 @@ class test_ucp_wireup_fallback_amo : public test_ucp_wireup {
+    ASSERT_UCS_OK(status);
+    void *req = ucp_ep_flush_nb(sender().ep(), UCT_FLUSH_FLAG_LOCAL, flush_cb);
+    if (UCS_PTR_IS_PTR(req)) {
+        test_ucp_wireup_amo **req_headroom = (test_ucp_wireup_amo **)
```
seems wrong, the user data is after UCP, not before, this seems to actually override ucp_request_t portion

yosefe commented on test/gtest/ucp/test_ucp_wireup.cc:
```diff
@@ -1149,6 +1205,60 @@ class test_ucp_wireup_fallback_amo : public test_ucp_wireup {
+
+    ucs_status_t status = ucp_atomic_post(sender().ep(), UCP_ATOMIC_POST_OP_ADD,
+                                          get_amo_value(), sizeof(elem_type),
+                                          get_amo_recv_addr(), rkey);
```
no need the wrapper get_amo_recv_addr

yosefe commented on test/gtest/ucp/test_ucp_wireup.cc:
```diff
@@ -1149,6 +1205,60 @@ class test_ucp_wireup_fallback_amo : public test_ucp_wireup {
+    static ucp_params_t get_ctx_params() {
+        ucp_params_t params = test_ucp_wireup::get_ctx_params();
+        params.field_mask  |= UCP_PARAM_FIELD_REQUEST_SIZE;
+        params.request_size = sizeof(test_ucp_wireup_amo *);
```
better define struct for request

yosefe commented on test/gtest/ucp/test_ucp_wireup.cc:
```diff
@@ -200,17 +221,52 @@ ucp_rkey_h test_ucp_wireup::get_rkey(ucp_ep_h ep, ucp_mem_h memh)
 }
 
+void test_ucp_wireup::fill_send_data() {
+    std::fill(m_send_data.begin(), m_send_data.end(), ucs_generate_uuid(0));
```
- move this function to test_ucp_wireup_amo 
- set only one var, not an array

yosefe commented on test/gtest/ucp/test_ucp_wireup.cc:
```diff
@@ -200,17 +221,52 @@ ucp_rkey_h test_ucp_wireup::get_rkey(ucp_ep_h ep, ucp_mem_h memh)
     m_memh_receiver.reset();
+}
+
+test_ucp_wireup::elem_type test_ucp_wireup::get_amo_value() const {
```
remove this wrapper

yosefe commented on test/gtest/ucp/test_ucp_wireup.cc:
```diff
@@ -68,15 +69,29 @@ class test_ucp_wireup : public ucp_test {
     static void tag_recv_completion(void *request, ucs_status_t status,
                                     ucp_tag_recv_info_t *info);
 
+    ucp_rkey_h get_rkey(const entity &e);
```
move all functions not used by test_ucp_wireup to test_ucp_wireup_amo class

evgeny-leksikov commented:
rpmbuild -bs '--define=_rpmdir /scrap/jenkins/workspace/ucx/build-test/rpm-dist' '--define=_srcrpmdir /scrap/jenkins/workspace/ucx/build-test/rpm-dist' '--define=_sourcedir /scrap/jenkins/workspace/ucx/build-test' '--define=_specdir /scrap/jenkins/workspace/ucx/build-test' '--define=_builddir /scrap/jenkins/workspace/ucx/build-test' --buildroot=/scrap/jenkins/workspace/ucx/build-test/_rpm --nodeps ucx.spec

[2019-12-17T07:27:42.660Z] error: Bad owner/group: /scrap/jenkins/workspace/ucx/build-test/ucx-1.8.0.tar.gz

bot:retest

mellanox-github commented:
<details><summary>Mellanox CI: <b>PASSED</b#  > on 25 workers (click for details)</summary>

**Note:** the logs will be deleted after 27-Dec-2019

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/b0fe8f1211671c51454e76ff4afe7689/raw/016f8c6e57a83c16b739977047fb6e2a41f6f935/_main) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/b0fe8f1211671c51454e76ff4afe7689/raw/8fe042ea44961cb671b012b749df0a992e09395f/hpc-arm-cavium-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/b0fe8f1211671c51454e76ff4afe7689/raw/26c731cd10d30f4f774fad77874c3df8b7a61bc4/hpc-arm-cavium-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/b0fe8f1211671c51454e76ff4afe7689/raw/245852a420e18e6e136cb6e481b23d851d08b06b/hpc-arm-cavium-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/b0fe8f1211671c51454e76ff4afe7689/raw/38a89c8061683a74fc2f822b1c38179388428143/hpc-arm-cavium-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/b0fe8f1211671c51454e76ff4afe7689/raw/93165a3ac6aa579efc59738fb1f24e77c6ac0541/hpc-arm-hwi-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/b0fe8f1211671c51454e76ff4afe7689/raw/86ebbba1e574c434a8776c8c9cd3be6f8417e064/hpc-arm-hwi-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/b0fe8f1211671c51454e76ff4afe7689/raw/46f3a3e87138ca334daa31cae560315c4ad8d2ef/hpc-arm-hwi-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/b0fe8f1211671c51454e76ff4afe7689/raw/f072a3144ff3c69df27e4c549d80e17e901d6256/hpc-arm-hwi-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W0](https://gist.githubusercontent.com/mellanox-github/b0fe8f1211671c51454e76ff4afe7689/raw/a0d078452a05c16f2e3e05a60489ca0c47240b0e/hpc-test-node-gpu_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W1](https://gist.githubusercontent.com/mellanox-github/b0fe8f1211671c51454e76ff4afe7689/raw/de326dc7db730ede044b3af84a1463d313c9d1d8/hpc-test-node-gpu_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W2](https://gist.githubusercontent.com/mellanox-github/b0fe8f1211671c51454e76ff4afe7689/raw/7a6e259ad0c107f0f4fda5b9bf8d89de0e345397/hpc-test-node-gpu_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W3](https://gist.githubusercontent.com/mellanox-github/b0fe8f1211671c51454e76ff4afe7689/raw/e47fcc0bba0cad72ebd5c98c172a0467f3cdee5b/hpc-test-node-gpu_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W0](https://gist.githubusercontent.com/mellanox-github/b0fe8f1211671c51454e76ff4afe7689/raw/9faa5d61a81439353ef7eb315c36cb050e93d72e/hpc-test-node-legacy_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W1](https://gist.githubusercontent.com/mellanox-github/b0fe8f1211671c51454e76ff4afe7689/raw/a4cc2c7485eb0a6debf4426319fe9db0858c7cc2/hpc-test-node-legacy_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W2](https://gist.githubusercontent.com/mellanox-github/b0fe8f1211671c51454e76ff4afe7689/raw/0082ee98468e750c37f0fa06029fb6321cd6fb01/hpc-test-node-legacy_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W3](https://gist.githubusercontent.com/mellanox-github/b0fe8f1211671c51454e76ff4afe7689/raw/e7914358d824b359f2708dafb1051909aab1df0a/hpc-test-node-legacy_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W0](https://gist.githubusercontent.com/mellanox-github/b0fe8f1211671c51454e76ff4afe7689/raw/79750f4d5fe8d2ed83a91e5e947abc666fd92353/hpc-test-node-new_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W1](https://gist.githubusercontent.com/mellanox-github/b0fe8f1211671c51454e76ff4afe7689/raw/56e9e5399787e7773a49bf55622f6947f4a092a1/hpc-test-node-new_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W2](https://gist.githubusercontent.com/mellanox-github/b0fe8f1211671c51454e76ff4afe7689/raw/1e8e9415a250d7266a02300eb2e5359169fd609a/hpc-test-node-new_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W3](https://gist.githubusercontent.com/mellanox-github/b0fe8f1211671c51454e76ff4afe7689/raw/ed2e36daa43e54a805fc5a3712a7b20a902cf97b/hpc-test-node-new_W3) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/b0fe8f1211671c51454e76ff4afe7689/raw/560b79781b79e7ce11e4b8bfd03939a0abedb41e/r-vmb-ppc-jenkins_W0) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/b0fe8f1211671c51454e76ff4afe7689/raw/8cdad840d49e6f271b41cbc40e3f5717ea76c47c/r-vmb-ppc-jenkins_W1) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/b0fe8f1211671c51454e76ff4afe7689/raw/6b98200bce879dee0d4110a51a314f9c46ffbd7a/r-vmb-ppc-jenkins_W2) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/b0fe8f1211671c51454e76ff4afe7689/raw/0f370d39fa3f38218066b38d929b7a9993e2f379/r-vmb-ppc-jenkins_W3) | :heavy_check_mark: SUCCESS
